p2p/src/authenticated/discovery/types.rs

```diff
@@ -160,8 +180,70 @@ pub struct PeerInfo<C: PublicKey> {
     pub signature: C::Signature,
 }
 
-impl<C: PublicKey> PeerInfo<C> {
-    /// Verify the signature of the peer info.
+/// Validate peer gossip payloads against configurability and basic safety checks.
+#[derive(Clone)]
+pub struct InfoVerifier<C: PublicKey> {
+    /// The [PublicKey] of the verifier.
+    me: C,
+
+    /// Whether to allow private IPs.
+    allow_private_ips: bool,
+
+    /// The maximum number of [Info] allowable in a single message.
+    peer_gossip_max_count: usize,
+
+    /// The time bound for synchrony. Messages with timestamps greater than this far into the
+    /// future will be considered malformed.
+    synchrony_bound: Duration,
+
+    /// The namespace used to sign and verify [Info] messages.
+    ip_namespace: Vec<u8>,
+}
+
+impl<C: PublicKey> InfoVerifier<C> {
+    /// Handle an incoming list of peer information.
+    ///
+    /// Returns an error if the list itself or any entries can be considered malformed.
+    pub fn validate(&self, clock: &impl Clock, infos: &[Info<C>]) -> Result<(), Error> {
+        // Ensure there aren't too many peers sent
+        if infos.len() > self.peer_gossip_max_count {
+            return Err(Error::TooManyPeers(infos.len()));
+        }
+
+        // We allow peers to be sent in any order when responding to a bit vector (allows
+        // for selecting a random subset of peers when there are too many) and allow
+        // for duplicates (no need to create an additional set to check this)
+        for info in infos {
+            // Check if IP is allowed
+            #[allow(unstable_name_collisions)]
+            if !self.allow_private_ips && !info.socket.ip().is_global() {
+                return Err(Error::PrivateIPsNotAllowed(info.socket.ip()));
+            }
+
+            // Check if peer is us
+            if info.public_key == self.me {
+                return Err(Error::ReceivedSelf);
+            }
+
+            // If any timestamp is too far into the future, disconnect from the peer
```

Comment:
nit: here and on line 235, remove "disconnect from the peer" since this function does not do that directly (it only returns an error)

Reply:
fair